File: 0012-integer-to-roman/0012-integer-to-roman.cs
public class Solution {
    public string IntToRoman(int num) {
        var dict = new Dictionary<int, string>(){
            {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"}, {100, "C"}, 
            {90, "XC"}, {50, "L"}, {40, "XL"}, {10, "X"}, 
            {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}
        };

        var ans = new StringBuilder();
        foreach(var (decim, roman) in dict){
            if(num == 0) break;
            while(num >= decim){
                ans.Append(roman);
                num -= decim;
            }
        }

        return ans.ToString();
    }
}

// Accepted - O(1)
public class SolutionQuick {
    public string IntToRoman(int num) {
        var M = new string[]{"", "M", "MM", "MMM"};
        var C = new string[]{"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
        var X = new string[]{"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
        var I = new string[]{"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};
        return M[num/1000] + C[(num%1000)/100] + X[(num%100)/10] + I[num%10];
    }
}

File: 0012-integer-to-roman/0012-integer-to-roman.java
class Solution {
    public String intToRoman(int num) {
        var map = GetNumberToRomanMap();

        var sb = new StringBuilder();
        for(var pair : map.entrySet()){
            while(num >= pair.getKey()){
                sb.append(pair.getValue());
                num -= pair.getKey();
            }
        }

        return sb.toString();
    }

    private Map<Integer, String> GetNumberToRomanMap(){
        var map = new LinkedHashMap<Integer, String>();
        map.put(1000, "M");
        map.put(900, "CM");
        map.put(500, "D");
        map.put(400, "CD");
        map.put(100, "C");
        map.put(90, "XC");
        map.put(50, "L");
        map.put(40, "XL");
        map.put(10, "X");
        map.put(9, "IX");
        map.put(5, "V");
        map.put(4, "IV");
        map.put(1, "I");

        return map;
    }
}